src/engine.rs

```diff
@@ -113,74 +110,64 @@ where
 #[derive(Clone)]
 pub struct Engine<E, W, P>
 where
-    E: Message,
+    E: Message + Clone,
     W: EntryExt<E>,
-    P: PipeLog,
+    P: GenericPipeLog,
 {
     cfg: Arc<Config>,
     memtables: MemTableAccessor<E, W>,
     pipe_log: P,
     cache_stats: Arc<SharedCacheStats>,
+    purge_manager: PurgeManager<E, W, P>,
 
     workers: Arc<RwLock<Workers>>,
-
-    // To protect concurrent calls of `purge_expired_files`.
-    purge_mutex: Arc<Mutex<()>>,
 }
 
 impl<E, W, P> Engine<E, W, P>
 where
     E: Message + Clone,
     W: EntryExt<E> + 'static,
-    P: PipeLog,
+    P: GenericPipeLog,
 {
-    // recover from disk.
+    // Recover from disk.
     fn recover(
-        pipe_log: &mut P,
+        queue: LogQueue,
+        pipe_log: &P,
         memtables: &MemTableAccessor<E, W>,
         recovery_mode: RecoveryMode,
     ) -> Result<()> {
         // Get first file number and last file number.
-        let first_file_num = pipe_log.first_file_num();
-        let active_file_num = pipe_log.active_file_num();
+        let first_file_num = pipe_log.first_file_num(queue);
+        let active_file_num = pipe_log.active_file_num(queue);
 
+        // Iterate and recover from files one by one.
         let start = Instant::now();
+        for file_num in first_file_num..=active_file_num {
+            // Read a file.
+            let content = pipe_log.read_whole_file(queue, file_num)?;
 
-        // Iterate files one by one
-        let mut current_read_file = first_file_num;
-        while current_read_file <= active_file_num {
-            // Read a file
-            let content = pipe_log
-                .read_next_file()
-                .unwrap_or_else(|e| {
-                    panic!(
-                        "Read content of file {} failed, error {:?}",
-                        current_read_file, e
-                    )
-                })
-                .unwrap_or_else(|| panic!("Expect has content, but get None"));
-
-            // Verify file header
+            // Verify file header.
             let mut buf = content.as_slice();
             if buf.len() < FILE_MAGIC_HEADER.len() || !buf.starts_with(FILE_MAGIC_HEADER) {
-                if current_read_file != active_file_num {
-                    panic!("Raft log file {} is corrupted.", current_read_file);
+                if file_num != active_file_num {
+                    warn!("Raft log header is corrupted at {:?}.{}", queue, file_num);
+                    return Err(box_err!("Raft log file header is corrupted"));
```

Comment:
Would the corruption happens just after the file has been created but not written HEADER?

Reply:
In this case `file_num == active_file_num` will be true.